internal/fs/kernel_list_cache_test.go

```diff
@@ -0,0 +1,353 @@
+	assert.Nil(t.T(), err)
+	assert.Equal(t.T(), 2, len(names1))
+	assert.Equal(t.T(), names1[0], "file1.txt")
+	assert.Equal(t.T(), names1[1], "file2.txt")
+	err = f.Close()
+	assert.Nil(t.T(), err)
+	// Adding one object to make sure to change the ReadDir() response.
+	assert.Nil(t.T(), t.createObjects(map[string]string{
+		"explicitDir/file3.txt": "123456",
+	}))
+	defer t.deleteObjectOrFail("explicitDir/file3.txt")
+	// Advancing the time more than ttl.
+	cacheClock.AdvanceTime(kernelListCacheTtlSeconds*time.Second + time.Second)
+
+	// Since out of ttl, so invalidation happens and ReadDir() will be served from
+	// gcsfuse filesystem.
+	f, err = os.Open(path.Join(mntDir, "explicitDir"))
+	assert.Nil(t.T(), err)
+	names2, err := f.Readdirnames(-1)
+
+	assert.Nil(t.T(), err)
+	assert.Equal(t.T(), 3, len(names2))
+	assert.Equal(t.T(), names2[0], "file1.txt")
+	assert.Equal(t.T(), names2[1], "file2.txt")
+	assert.Equal(t.T(), names2[2], "file3.txt")
+}
+
+// TestKernelListCache_CacheHitAfterInvalidation:
+// (a) First read will be served from GcsFuse filesystem.
+// (b) Second read after ttl will also be served from GCSFuse file-system.
+// (c) Third read within ttl will be served from kernel page cache.
+func (t *KernelListCacheTestWithPositiveTtl) TestKernelListCache_CacheHitAfterInvalidation() {
+	// First read, kernel will cache the dir response.
+	f, err := os.Open(path.Join(mntDir, "explicitDir"))
+	assert.Nil(t.T(), err)
+	defer func() {
+		assert.Nil(t.T(), f.Close())
+	}()
+	names1, err := f.Readdirnames(-1)
+	assert.Nil(t.T(), err)
+	assert.Equal(t.T(), 2, len(names1))
+	assert.Equal(t.T(), names1[0], "file1.txt")
+	assert.Equal(t.T(), names1[1], "file2.txt")
+	err = f.Close()
+	assert.Nil(t.T(), err)
+	// Adding one object to make sure to change the ReadDir() response.
+	assert.Nil(t.T(), t.createObjects(map[string]string{
+		"explicitDir/file3.txt": "123456",
+	}))
+	defer t.deleteObjectOrFail("explicitDir/file3.txt")
+	// Advancing the time more than ttl.
+	cacheClock.AdvanceTime(kernelListCacheTtlSeconds*time.Second + time.Second)
+	// Since out of ttl, so invalidation happens and ReadDir() will be served from
+	// gcsfuse filesystem.
+	f, err = os.Open(path.Join(mntDir, "explicitDir"))
+	assert.Nil(t.T(), err)
+	names2, err := f.Readdirnames(-1)
+	assert.Nil(t.T(), err)
+	assert.Equal(t.T(), 3, len(names2))
+	assert.Equal(t.T(), names2[0], "file1.txt")
+	assert.Equal(t.T(), names2[1], "file2.txt")
+	assert.Equal(t.T(), names2[2], "file3.txt")
+	err = f.Close()
+	assert.Nil(t.T(), err)
+	// Adding one object to make sure to change the ReadDir() response.
+	assert.Nil(t.T(), t.createObjects(map[string]string{
+		"explicitDir/file4.txt": "123456",
+	}))
+	defer t.deleteObjectOrFail("explicitDir/file4.txt")
+	// Advancing the time within ttl.
+	cacheClock.AdvanceTime(kernelListCacheTtlSeconds * time.Second / 2)
+
+	// Within ttl, so will be served from kernel.
+	f, err = os.Open(path.Join(mntDir, "explicitDir"))
```

Comment:
Have we closed this f? same for all the other places

Reply:
Yes, please check the defer statement. The last `f` is close as part of defer statement.